versiondb/client/convert_to_sst.go

```diff
@@ -0,0 +1,223 @@
+	cmd.Flags().String(flagStore, "", "store name, the keys are prefixed with \"s/k:{store}/\"")
+	cmd.Flags().Int64(flagSorterChunkSize, DefaultSorterChunkSize, "uncompressed chunk size for external sorter, it decides the peak ram usage, on disk it'll be snappy compressed")
+	return cmd
+}
+
+// sstFileName inserts the seq integer into the base file name
+func sstFileName(fileName string, seq int) string {
+	stem := fileName[:len(fileName)-len(SSTFileExtension)]
+	return stem + fmt.Sprintf("-%d", seq) + SSTFileExtension
+}
+
+func newSSTFileWriter() *grocksdb.SSTFileWriter {
+	envOpts := grocksdb.NewDefaultEnvOptions()
+	return grocksdb.NewSSTFileWriter(envOpts, tsrocksdb.NewVersionDBOpts(true))
+}
+
+// encodeSorterItem encode kv-pair for use in external sorter.
+//
+// layout: key + version(8) + delete(1) + [ value ] + key length(SizeKeyLength)
+// we put the key and version in the front of payload so it can take advantage of the delta encoding in the `ExtSorter`.
+func encodeSorterItem(version uint64, pair iavl.KVPair) []byte {
+	item := make([]byte, sizeOfSorterItem(pair))
+	copy(item, pair.Key)
+	offset := len(pair.Key)
+
+	binary.LittleEndian.PutUint64(item[offset:], version)
+	offset += 8
+
+	if pair.Delete {
+		item[offset] = 1
+		offset++
+	} else {
+		copy(item[offset+1:], pair.Value)
+		offset += len(pair.Value) + 1
+	}
+	binary.LittleEndian.PutUint32(item[offset:], uint32(len(pair.Key)))
+	return item
+}
+
+// sizeOfSorterItem compute the encoded size of pair
+//
+// see godoc of `encodeSorterItem` for layout
+func sizeOfSorterItem(pair iavl.KVPair) int {
+	size := len(pair.Key) + 8 + 1 + SizeKeyLength
+	if !pair.Delete {
+		size += len(pair.Value)
+	}
+	return size
+}
+
+// decodeSorterItem decode the kv-pair from external sorter.
+//
+// see godoc of `encodeSorterItem` for layout
+func decodeSorterItem(item []byte) ([]byte, iavl.KVPair) {
+	var value []byte
+	keyLen := binary.LittleEndian.Uint32(item[len(item)-SizeKeyLength:])
+	key := item[:keyLen]
+	version := item[keyLen : keyLen+8]
+	delete_ := item[keyLen+8] == 1
+	if !delete_ {
+		value = item[keyLen+9 : len(item)-SizeKeyLength]
+	}
+	return version, iavl.KVPair{
+		Delete: delete_,
+		Key:    key,
+		Value:  value,
+	}
+}
+
+// compareSorterItem compare encoded kv-pairs return if a < b.
+func compareSorterItem(a, b []byte) bool {
+	// decode key and version
+	aKeyLen := binary.LittleEndian.Uint32(a[len(a)-SizeKeyLength:])
+	bKeyLen := binary.LittleEndian.Uint32(b[len(b)-SizeKeyLength:])
+	ret := bytes.Compare(a[:aKeyLen], b[:bKeyLen])
+	if ret != 0 {
+		return ret == -1
+	}
+
+	aVersion := binary.LittleEndian.Uint64(a[aKeyLen:])
+	bVersion := binary.LittleEndian.Uint64(b[bKeyLen:])
+	// Compare version.
+	// For the same user key with different timestamps, larger (newer) timestamp
+	// comes first.
+	return aVersion > bVersion
+}
+
+func cloneAppend(bz []byte, tail []byte) (res []byte) {
```

Comment:
maybe can [cloneAppend](https://github.com/crypto-org-chain/cronos/blob/4900fe0333c7b70c18d18a39fda997e402adb5bc/versiondb/tsrocksdb/store.go#L195) reuse to util, but not sure if need clone since no mutate in both places

Reply:
`append` itself can mutate the first parameter if there's capacity.